internal/interpolate/parse_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interpolate
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseSuccess(t *testing.T) {
+	tests := []struct {
+		give string
+		want String
+	}{
+		{
+			give: "foo",
+			want: String{literal("foo")},
+		},
+		{
+			give: "foo ${bar} baz",
+			want: String{
+				literal("foo "),
+				variable{Name: "bar"},
+				literal(" baz"),
+			},
+		},
+		{
+			give: "foo $ {bar} baz",
+			want: String{literal("foo "), literal("$ "), literal("{bar} baz")},
+		},
+		{
+			give: "foo ${bar:}",
+			want: String{
+				literal("foo "),
+				variable{Name: "bar", HasDefault: true},
+			},
+		},
+		{
+			give: "${foo:bar}",
+			want: String{
+				variable{
+					Name:       "foo",
+					Default:    "bar",
+					HasDefault: true,
+				},
+			},
+		},
+		{
+			give: `foo \${bar:42} baz`,
+			want: String{
+				literal("foo "),
+				literal("$"),
+				literal("{bar:42} baz"),
+			},
+		},
+		{
+			give: "$foo${bar}",
+			want: String{
+				literal("$f"),
+				literal("oo"),
+				variable{Name: "bar"},
+			},
+		},
+		{
+			give: "foo${b-a-r}",
+			want: String{
+				literal("foo"),
+				variable{Name: "b-a-r"},
+			},
+		},
+		{
+			give: "foo ${bar::baz} qux",
```

Comment:
Already added a test for it. Variable names cannot contain whitespace. Default values can (which I just realized I don't have a test for; adding0.